### PR TITLE
New formats for integration test given-when-then step text

### PR DESCRIPTION
**Pull request by phil-davis**
## Description
Change the "integration" (=API) gherkin test steps to a new "standard" format.
Refactor the code behind to do what the steps say - adding new methods etc to handle the extra variations in possible step text.
Behat Gherkin is able to match any of multiple ``Given`` ``When`` annotations in the PHPdoc for a single method, so make use of that when there are different wordings for a step which have the same parameter order. 

## Related Issue
https://github.com/owncloud/QA/issues/517

## Motivation and Context
Make the integration-API test steps have a consistent agreed style. Then we can move on to analysing the test coverage, effectiveness, merging in UI tests, renaming the "integration" folder to "acceptance" and so on...

## How Has This Been Tested?
CI passes

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Refactoring (tests only)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by phil-davis** [comment]
This is a "squashed into 2 commits" version of #30233 
(I wanted to keep history of bits and pieces I had done along the way in that PR, in case I wonder)

**Comment by phil-davis** [comment]
The changes here implement:
1) Role vs actual user name
```
And as an "admin"
```
becomes:
```
And as user "admin"
```
The old way looks like it might be specifying a role, like "as an accountant". The new way is clearer that it is just some (arbitrary) user name.

2) Remove ``is`` ``are`` ``exists`` etc from steps that are intended to be a ``Given`` that sets up state:
```
Given user "user0" exists
```
becomes
```
Given user "user0" has been created
```
this reduces confusion with ``Then`` steps that test state - ``Then user "user0" exists`` could be a step that expects the user to already exist, and just test for it.

3) Remove ``is`` ``are`` ``exists`` etc from steps that are intended to be a ``Then`` that tests state:
```
Then as "user1" the file "/local.txt" exists
```
becomes
```
Then as "user1" the file "/local.txt" should exist
```
The word ``should`` appears somewhere in every ``Then`` step - that makes it easy to understand that the step will test for something.

4) Write ``Given`` steps using present perfect tense, saying what "has been done":
```
Given user "user0" has created a folder "/local_storage/foo2"
```
for ``Given`` steps, we do not care how "the system" achieves the ``Given`` (by doing some ``occ`` command magic, hitting the testing app, some API requests, or driving the webUI). The objective is just that "the system under test" is brought to the ``Given`` state.

5) Write ``When`` steps in the active present tense - some subject (user, client, administrator...) does something, and by what mechanism:
```
When user "user1" moves file "/local_storage/foo2/textfile0.txt" to "/local.txt" using the API
```
for ``When`` we want some action-under-test to happen, so we care to know "how" the action is being done. e.g. we are testing the API ``using the API`` or testing the command line ``using the occ command`` and (when UI tests appear here) ``using the webUI``

**Comment by phil-davis** [comment]
This is ready for review. I am hoping it provides a starting point for then moving forward with items like:
- [ ] update the UI test gherkin to be like this
- [ ] rename the ``integration`` folder to ``acceptance``
- [ ] put all the API and UI tests "together" into the ``acceptance`` folder (in appropriately designed suites)
- [ ] review the existing tests and their coverage of the API and UI and decide what areas to address first to fill coverage holes
...

**Comment by phil-davis** [comment]
Before merging:
- [ ] review integration tests in apps and make sure/make them work with this core test code

For now I kept many of the "old" ``core`` methods in the code, tagging them ``deprecated`` in the PHPdoc - that will help minimize the "absolutely essential" changes that have to be made to integration test feature code in apps. After fixing up all app integration tests, we can then remove everything tagged ``deprecated``.

File: tests/integration/features/bootstrap/BasicStructure.php [hunk 1]
```diff
@@ -95,16 +95,25 @@ private function baseUrlWithoutOCSAppendix() {
 	}
 
 	/**
-	 * @Given /^using api version "([^"]*)"$/
+	 * @Given /^using (?:api|API) version "([^"]*)"$/
```
Comment (individual-it) [inline tests/integration/features/bootstrap/BasicStructure.php]: would `/i` at the end of the regex work?

**Review by individual-it — APPROVED** [review]
a lot of good stuff and CI passes, so it must be useful 👍

File: tests/integration/features/bootstrap/BasicStructure.php (hunk 1 above)
Comment (phil-davis) [reply tests/integration/features/bootstrap/BasicStructure.php]: probably - then it would let you say "UsInG" and "aPi" and "VeRsIoN" ...

Comment (individual-it) [reply tests/integration/features/bootstrap/BasicStructure.php]: that is not camelCase but KaNgArOoCaSe

**Comment by phil-davis** [comment]
DO NOT MERGE YET.
To do:
- [ ] get more feedback-review from others that this is "a good thing"

Review integration tests in the following apps and have PRs ready for any core gherkin test steps that need to be changed to work with this refactored code:
- [ ] customgroups
- [ ] firewall
- [ ] guests
- [ ] notifications
- [ ] objectstore
- [ ] ransomware_protection
- [ ] user_ldap
- [ ] windows_network_drive

That is all the apps I can see that have Behat-Gherkin style integration tests.

**Comment by PVince81** [comment]
maybe to avoid breaking apps, have some legacy steps/regexp and tag these as such ?

**Comment by phil-davis** [comment]
It is quite easy to update the test steps in the apps - I am already on a roll with PHP storm, and since 9.* is EOL we do not have to think about app integration tests still running against core 9.* etc. So this is a point at which we can sync the merging of all the refactorings (core master, core stable10 and apps).

Then there are not deprecated step formats hanging around that people might use.

**Comment by phil-davis** [comment]
DO NOT MERGE YET.
To do:
- [ ] get feedback-review on the PRs for app integration test refactoring

Please have a look at those (see links in post above) and give comments or approval.

**Comment by phil-davis** [comment]
Backport stable10 #30488

**Comment by phil-davis** [comment]
Note to self: just awaiting approval on WND PR. That is all that remains to be merged.

**Pull request merged**